README.md
# exportWindowsSources: Export Windows Sources from Active Directory for import into NetWitness

## Version 0.1

*exportWindowsSources* is a Windows PowerShell tool which allows one to export enabled Windows Event sources from Active Directory, in bulk.  It outputs CSV text which can be either directly imported to a Log Collector through Kensington Tech's AddWindowsSources automated solution, or manually via the NetWitness user interface.

## IMPORTANT NOTE

This script returns only computer objects which are enabled on the target domain.  Disabled computers will not be displayed.


## Prerequisites
1.  A Windows host with PowerShell 2 or higher installed

2.  The 'Remote Server Administration Tools' Windows feature must be installed, and the 'Active Directory module for Windows Powershell' module of said feature must be enabled.  For more info, visit https://4sysops.com/archives/how-to-install-the-powershell-active-directory-module

3.  An Active Directory account with sufficient permissions to access Active Directory Computer objects for a given AD domain.  This user can run exportWindowsSources either interactively (through PowerShell itself) or as a service account through the Windows scheduler.


## Installation
1.  Copy the exportWindowsSources.ps1 script to the Windows host that you've chosen, placing it in a directory which the user has permission to access.


## Usage

Usage: `$($scriptName) domainSuffix [ [DomainSearchBase] [DomainControllerHost] ]`

Option | Description
------ | -----------
domainSuffix | (required) is the DNS suffix for computer objects which do not possess the DNSHostName attribute, so that the Name attribute can be fully-qualified (i.e. a computer called 'mycomputer' can be named 'mycomputer.subdomain.mydomain.com').  This typically should be the domain suffix of the domain you're fetching from (e.g. 'mysubdomain.mydomain.com').  Do not include a leading period
domainSearchBase | (optional) is LDAP search DN of the domain to query, like "DC=MyDomain,DC=com".  If specified, DomainControllerHost must also be specified
domainControllerHost | (optional) is the hostname of a domain controller for the domain you're accessing

## Examples

Runs on current AD domain (mydomain.com):   `& ".\exportWindowsSources.ps1" mydomain.com`
Runs on AD domain mydomain.com:             `& ".\exportWindowsSources.ps1" mydomain.com "DC=mydomain,DC=com" mydc.mydomain.com`
Runs on AD domain subdomain.mydomain.com:   `& ".\exportWindowsSources.ps1" subdomain.mydomain.com "DC=subdomain,DC=mydomain,DC=com" mydc.subdomain.mydomain.com`


## SAVING FILES TO DISK
Similar to many UNIX tools, this tool DOES NOT write files directly to disk.  By default, it outputs to the local console.  To write the CSV to disk, its output must be redirected to a file using the '>' (overwrite) or '>>' (append) operators.

E.g.

`& ".\exportWindowsSources.ps1" mydomain.com "DC=mydomain,DC=com" mydc.mydomain.com > mydomain.com_export.csv`


## Troubleshooting

* If Windows refuses to execute the script, please know that this script is distributed in an unsigned form, therefore the Powershell execution policy must be set to 'Bypass' (no warning displayed) or 'Unrestricted' (displays a warning first) to run it.  To check the policy, run 'Get-ExecutionPolicy' from Powershell.  For more info, visit https://technet.microsoft.com/en-us/library/ee176961.aspx

* If the script outputs empty CSV files, it will display an error on the command line.  If you are running the tool as a scheduled task, it may be necessary to run the command interactively with Windows Powershell using the service account of the scheduled task.  To do this, run Powershell, and type the command:

`start powershell -credential ""`


## Changing Default CSV Values
If it should be necessary for your environment to change the default values which are output by the tool, the script file itself must be edited.  Change the value of the below variables to suit your neeeds:

Variable | Description
-------- | -----------
$portNumber | Usually either 5985 (for HTTP) or 5986 (for HTTPS).  Defaults to 5985
$transportMode | Either 'http' or 'https'.  Defaults to 'http'
$debug | Either '0' (disabled - default) or '1' (enabled).  Enables debug logging in Log Collector logs for the event sources
$enabled | Either 'true' (default) or 'false'.  Enables or disables the event source.  Defaults to 'true'
$certName | The name of the certificate to use to authenticate to WinRM.  Defaults to an empty string ('').  Specified only if using HTTPS transport.  
$validateServer | Not documented by RSA
$renderLocale | The locale to render the event logs in.  Defaults to an empty string ('')
$resolveSids | Either 'true' (default) or 'false'.  Resolve Windows SID's to readable names
$sidsInterval | The number of seconds??? to cache resolved SID names before refreshing.  Defaults to 14400
$sidsTimeout | SID timeout, in seconds.  Defaults to 60
$overrideChannels | Defaults to an empty string ('')


## CSV Output Format and example
```
eventsource_address,port_number,transport_mode,debug,enabled,cert_name,validate_server,render_locale,windows_type,resolve_sids,sids_interval,sids_timeout,override_channels
dc1.mydomain.local,5985,http,0,true,,false,en-US,Domain Controller,true,14400,60,
dc2.mydomain.local,5985,http,0,true,,false,en-US,Domain Controller,true,14400,60,
host1.mydomain.local,5985,http,0,true,,false,en-US,Non-Domain Controller,true,14400,60,
```

Changelog
Version 0.1
16 July, 2017
Initial release

exportWindowsSources.ps1
### 
### exportWindowsSources
### Version 0.1
### 16 July, 2017
### Kensington Technology Associates, Limited
### info@knowledgekta.com
### http://knowledgekta.com
### 
### MIT License
### 
### Copyright (c) 2017, Kensington Technology Associates, Limited
### 
### Permission is hereby granted, free of charge, to any person obtaining a copy
### of this software and associated documentation files (the "Software"), to deal
### in the Software without restriction, including without limitation the rights
### to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
### copies of the Software, and to permit persons to whom the Software is
### furnished to do so, subject to the following conditions:
### 
### The above copyright notice and this permission notice shall be included in all
### copies or substantial portions of the Software.
### 
### THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
### IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
### FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
### AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
### LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
### OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
### SOFTWARE.
### 

Try {
  Import-Module ActiveDirectory -ErrorAction Stop
}
Catch {
  Write-Host "ERROR: Could not import the ActiveDirectory module for PowerShell.  Have you installed the 'Remote Server Administration Tools' Windows feature, and if so, have you enabled 'Active Directory module for Windows Powershell'?  See README for more info"
  exit
}

$version = '0.1'

$portNumber = '5985'
$transportMode = 'http'
$debug = '0'
$enabled = 'true'
$certName = ''
$validateServer = 'false'
$renderLocale = ''
$resolveSids = 'true'
$sidsInterval = '14400'
$sidsTimeout = '60'
$overrideChannels = ''


$scriptName = $MyInvocation.MyCommand.Name
$cpr = [char]0x00A9
$usage = @"
exportWindowsSources tool version $($version)
Copyright $($cpr) 2017, Kensington Technology Associates, Limited
Licensed under MIT License (See LICENSE file)

Usage: $($scriptName) domainSuffix [ [domainSearchBase] [domainControllerHost] ]

Outputs CSV notation describing enabled computers in a Windows AD domain.  This notation is in a format which can be imported into a NetWitness Log Collector automatically by Kensington Tech's addWindowsSources tool or manually via the RSA NetWitness console

Options:

domainSuffix - (required) is the DNS suffix for computer objects which do not possess the DNSHostName attribute, so that the Name attribute can be fully-qualified (i.e. a computer called 'mycomputer' can be named 'mycomputer.subdomain.mydomain.com'). This typically should be the domain suffix of the domain you're fetching from (e.g. 'mysubdomain.mydomain.com'). Do not include a leading period

domainSearchBase - (optional) is LDAP search DN of the domain to query, like 'DC=MyDomain,DC=com'. If specified, DomainControllerHost must also be specified

domainControllerHost - (optional) is the hostname of a domain controller for the domain you're accessing

"@

$sysArgsLen = $args.Length

if ($sysArgsLen -ne 1 -and $sysArgsLen -ne 3) {
  Write-Host $usage
  exit
}

$domainSuffix = $args[0]

$searchBase = ''
$domainController = ''
if ($args.length -eq 3) {
  $searchBase = $args[1]
  $domainController = $args[2]
}

function getSystems {
  $LDAPfilter = $args[0]
  #Write-Host $LDAPfilter
  $systems = @()
 
  Try {
   
    if ($sysArgsLen -eq 1) {
      $computers = Get-ADComputer -LDAPfilter $LDAPfilter -Properties 'DNSHostName','Name' -ErrorAction Stop
    }
    elseif ($sysArgsLen -eq 3) {
      $computers = Get-ADComputer -LDAPfilter $LDAPfilter -Properties 'DNSHostName','Name' -SearchBase $searchBase -Server $domainController -ErrorAction Stop
    }
    
    ForEach ($c in $computers) {
      if ( [bool]$c.DNSHostName -eq $True ) {
        $systems += $c.DNSHostName
      }
      elseif ([bool]$c.Name -eq $True) {
        $systems += "$($c.Name).$($domainSuffix)"
      }
    }
    return $systems
  }
  Catch {
    #$e = $_.Exception.GetType().FullName
    $e = $_.Exception.Message
    Write-Host "ERROR: $($e)"
    exit
  }
  
  
}

$dcFilter = "(&(objectCategory=computer)(userAccountControl:1.2.840.113556.1.4.803:=8192)(!userAccountControl:1.2.840.113556.1.4.803:=2))"
$nonDCFilter = "(&(objectCategory=Computer)(!userAccountControl:1.2.840.113556.1.4.803:=8192)(!userAccountControl:1.2.840.113556.1.4.803:=2))"

$nonDCs = getSystems $nonDCFilter
$DCs = getSystems $dcFilter

#print header
echo "eventsource_address,port_number,transport_mode,debug,enabled,cert_name,validate_server,render_locale,windows_type,resolve_sids,sids_interval,sids_timeout,override_channels"

$windowsType = 'Non-Domain Controller'
ForEach ($h in $nonDCs) {
  #echo "$($h),5985,http,0,true,,false,,Non-Domain Controller,true,14400,60,"
  echo "$($h),$($portNumber),$($transportMode),$($debug),$($enabled),$($certName),$($validateServer),$($renderLocale),$($windowsType),$($resolveSids),$($sidsInterval),$($sidsTimeout),$($overrideChannels)"
}

$windowsType = 'Domain Controller'
ForEach ($h in $DCs) {
  #echo "$($h),5985,http,0,true,,false,,Domain Controller,true,14400,60,"
  echo "$($h),$($portNumber),$($transportMode),$($debug),$($enabled),$($certName),$($validateServer),$($renderLocale),$($windowsType),$($resolveSids),$($sidsInterval),$($sidsTimeout),$($overrideChannels)"
}
